src/axom/quest/readers/ProEReader.cpp

```diff
@@ -0,0 +1,149 @@
+// Copyright (c) 2017-2023, Lawrence Livermore National Security, LLC and
+// other Axom Project Developers. See the top-level LICENSE file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+#include "axom/quest/readers/ProEReader.hpp"
+
+// Axom includes
+#include "axom/core/utilities/Utilities.hpp"
+#include "axom/mint/mesh/CellTypes.hpp"
+#include "axom/slic/interface/slic.hpp"
+
+// C/C++ includes
+#include <fstream>
+#include <map>
+
+//------------------------------------------------------------------------------
+//      ProEReader Implementation
+//------------------------------------------------------------------------------
+namespace axom
+{
+namespace quest
+{
+ProEReader::ProEReader() : m_fileName(""), m_num_nodes(0), m_num_tets(0) { }
+
+//------------------------------------------------------------------------------
+ProEReader::~ProEReader() { this->clear(); }
+
+//------------------------------------------------------------------------------
+void ProEReader::clear()
+{
+  m_num_nodes = 0;
+  m_num_tets = 0;
+  m_num_unique_nodes = 0;
+  m_nodes.clear();
+}
+
+//------------------------------------------------------------------------------
+int ProEReader::read()
+{
+  constexpr int NUM_NODES_PER_TET = 4;
+  constexpr int NUM_COMPS_PER_NODE = 3;
+
+  std::string junk;
+  std::string id;
+  std::string tet_nodes[NUM_NODES_PER_TET];
+
+  struct coordinate
+  {
+    double comp[NUM_COMPS_PER_NODE];
+  } cur_coord;
+
+  std::ifstream ifs(m_fileName.c_str());
+
+  if(!ifs.is_open())
+  {
+    SLIC_WARNING("Cannot open the provided Pro/E file [" << m_fileName << "]");
+    return (-1);
+  }
+
+  // Remove any comments from header
+  while(ifs.peek() == '#')
+  {
+    std::getline(ifs, junk);
+  }
+
+  // Initialize number of nodes and tetrahedra.
+  // 4 nodes per tet, 3 components per node.
+  ifs >> m_num_unique_nodes >> m_num_tets;
+  m_num_nodes = m_num_tets * NUM_NODES_PER_TET;
+  m_nodes.reserve(m_num_nodes * NUM_COMPS_PER_NODE);
+
+  // Read nodes
+  std::map<std::string, coordinate> nodes;
+  for(int i = 0; i < m_num_unique_nodes; i++)
+  {
+    ifs >> id >> cur_coord.comp[0] >> cur_coord.comp[1] >> cur_coord.comp[2];
+    nodes[id] = {cur_coord.comp[0], cur_coord.comp[1], cur_coord.comp[2]};
+  }
+
+  // Initialize nodes
+  for(int i = 0; i < m_num_tets; i++)
+  {
+    ifs >> id >> tet_nodes[0] >> tet_nodes[1] >> tet_nodes[2] >> tet_nodes[3];
+
+    for(int j = 0; j < NUM_NODES_PER_TET; j++)
+    {
+      for(int k = 0; k < NUM_COMPS_PER_NODE; k++)
+      {
+        m_nodes.push_back(nodes[tet_nodes[j]].comp[k]);
+      }
+    }
+  }
```

Comment:
The way this is implemented, we're losing track of the mesh connectivity.
I.e. every tet has 4 unique vertices rather than reusing the shared vertices from the input.

A more concise representation would be to store an array of coordinate positions, of size `m_num_unique_nodes` and an array of coordinate references (integers) of size `m_num_tets * 4`.  (I.e. as provided by the input mesh).

Since tet meshes can be large, it's probably worth implementing it that way and passing it to the mint mesh that way too. We can do this in a follow-up PR if you'd prefer.

Reply:
> A more concise representation would be to store an array of coordinate positions, of size m_num_unique_nodes and an array of coordinate references (integers) of size m_num_tets * 4. (I.e. as provided by the input mesh).

Refactored the storage so `m_nodes` are the coordinates and `m_tets` are the `int` IDs of the nodes.